File: algorithms/missing_integer/missing_integer.py
def solution(A):
    indicator = Indicator()
    output = indicator.split(A, 100000)
    print(output.val)
    return output.val

def half(n) :
  return (n - (n % 2)) / 2

class Array:
  def __init__(self, low, big):
    self._array = []
    self.low = low
    self.big = big

  def __len__(self):
    return len(self._array)
  
  def add(self, val):
    self._array.append(val)
    if (val > self.big) :
      self.big = val
    if (val < self.low) :
      self.low = val
  
  def __iter__(self):
    return iter(self._array)

class Indicator:
  val = 1
  next = 2

  def update(self, val):
    if (val < self.val):
      return False
    if (self.next == val) :
      self.next += 1
      return False
    if (self.val == val) :
      self.val = self.next
      self.next += 1
      return False
    return True

  def split(self, array, max) :
    n = len(array)
    if (n == 0) :
      return self
    pivot = half(self.val + max)
  
    left = Array(pivot - 1, self.val)
    right = Array(max, pivot)
    
    for item in array :
      if(self.update(item)):
        (left if item < pivot else right).add(item)

    if (self.val > right.big or self.val < left.low) :
      return self
    elif (self.val >= right.low) :
      return self.split(right, right.big)
    elif (self.val > left.big) :
      return self
    else:
      self.split(left, left.big)
      if (self.val < right.low or len(right) == 0) :
        return self
      return self.split(right, right.big)


solution([1, 3, 6, 1, 2])
solution([-1, -3])
solution([60000, 50000, 9, 5, 6, 8, 4, 1000, 3, 2, 1])
solution([9, 5, 6, 8, 4, 3, 2, 1])
solution([20, 2, 17, 4, 15, 6, 13, 8, 11, 10, 9, 12, 7, 14, 5, 16, 3, 19, 1])

File: algorithms/missing_integer/missing_integer.js
function solution(A) {
  const data = { val: 1, next: 2 };
  const output = split(data, A, 1000000);
  console.log(output.val);
  return output.val;
}

function split(data, array, max) {
  const n = array.len || array.length;
  if (n === 0) {
    return data;
  }

  const pivot = half(data.val + max);

  const left = new Array(n);
  left.len = 0;
  left.low = pivot - 1;
  left.big = data.val;

  const right = new Array(n);
  right.len = 0;
  right.big = pivot;
  right.low = max;

  for (var i = 0; i < n; i++) {
    if (array[i] < data.val) continue;

    if (data.next === array[i]) {
      data.next++;
      continue;
    } else if (data.val === array[i]) {
      data.val = data.next;
      data.next++;
      continue;
    }
    append(array[i] < pivot ? left : right, array[i]);
  }

  if (data.val > right.big || data.val < left.low) {
    return data;
  } else if (data.val >= right.low) {
    return split(data, right, right.big);
  } else if (data.val > left.big) {
    return data;
  } else {
    data = split(data, left, left.big);
    if (data.val < right.low || right.length === 0) {
      return data;
    }
    return split(data, right, right.big);
  }
}

function half(n) {
  return (n - (n % 2)) / 2;
}

function append(array, val) {
  array[array.len] = val;
  array.len++;
  if (val > array.big) {
    array.big = val;
  }
  if (val < array.low) {
    array.low = val;
  }
}

solution([1, 3, 6, 1, 2]);
solution([-1, -3]);
solution([60000, 50000, 9, 5, 6, 8, 4, 1000, 3, 2, 1]);
solution([9, 5, 6, 8, 4, 3, 2, 1]);
solution([20, 2, 17, 4, 15, 6, 13, 8, 11, 10, 9, 12, 7, 14, 5, 16, 3, 19, 1]);
